### PR TITLE
(time-namespaced state) Only rehydrate state from url for namespaces not seen since page load.

### PR DESCRIPTION
**Pull request by bmd3k**
As desribed in detail in b/211018035 (sorry, Googlers only), there is tension between ngrx-managed state and state as described in the URL.

We can get in situations where user navigates back and forward in browser history and the state rehydrated from the URL overwrites ngrx-managed state. In the case of namespaced state it can often overwrite the WRONG state since the rehydration happens before namespace resolution is completed.

The solution we propose is to rehydrate state from the URL less frequently. We skip rehydration from URL if navigating to a namespace that we already have tracked in ngrx-managed state.

This approach will solve a bunch of navigation bugs in route-namespaced state but will also avoid some weird potential navigation bugs in the new time-namespaced state.

It does have some certain idiosyncracies of its own that we will have to accept as intended behavior. Consider if user has generated the following browser history and namespaces since loading the app:

* (1) {http://urlA?tagFilter=A, namespaceId: A}
* (2) {http://urlAprime?tagFilter=Aprime, namespaceId: A}
* (3) {http://urlB?tagFilter=B, namespaceId: B}

Then two behaviors we'd have to accept as 'Working as Intended':

1. If user jumps from state (3) to state (1) then 'tagFilter=A' will be ignored and the tag filter will be 'Aprime'. This seems to be justifiable since it fits into our mental model about how namespaces are supposed to behave according to go/tb-timespaced-state.

2. If user reloads the page (thus clearing in-memory state) and then jumps from state (3) to state (1) then 'tagFilter=A' will be honored and the tag filter will be 'A'. If they then navigate forward to state (2) then 'tagFilter=Aprime' will be ignored and the tag filter will remain 'A'. This might be a little harder to justify but I think it is a compromise that makes sense given how the system works.

### CODE REVIEW

**Comment by japie1235813** [comment]
In your second WAI scenario, when user reloads the page, I assume user reloads at sate 3, does the concept of  namespace B & A still exist? I mean, would reloading page means generating a new namespace?
(4) {http://urlB?tagFilter=B, namespaceId: C}
The url is the same as (3) but namespace B is cleared. Thus when user goes from (3) url to (1) url, do we still have namespace A or it is another namespace A'?

Similar question, if user goes from (3) to (2), then the tagFilter rehydrate from url will be `Aprime`. Goes from there when user navigate back to (1), tagFilter remains as `Aprime`

What if user navigate to (2) from (1) and then hit back navigation to (1), in this logic tagFilter should be `Aprime` but does that match users mental model?

**Comment by japie1235813** [comment]
To add to my previous comment, when user refreshes page at (3) we still want to remain in namespace B according to https://github.com/tensorflow/tensorboard/pull/5470?

**Comment by bmd3k** [comment]
> In your second WAI scenario, when user reloads the page, I assume user reloads at sate 3, does the concept of namespace B & A still exist? I mean, would reloading page means generating a new namespace? (4) {http://urlB?tagFilter=B, namespaceId: C} The url is the same as (3) but namespace B is cleared. Thus when user goes from (3) url to (1) url, do we still have namespace A or it is another namespace A'?

As you later discovered, #5470 makes it so that we remain in namespace B on page refresh at (3). We also continue to maintain namespace A if user navigates back to (1) and (2). One of the advantages of keeping namespaceId in browser history state is it allows us to best-effort honor the original namespaces even on page refresh.

> 
> Similar question, if user goes from (3) to (2), then the tagFilter rehydrate from url will be `Aprime`. Goes from there when user navigate back to (1), tagFilter remains as `Aprime`
> 
> What if user navigate to (2) from (1) and then hit back navigation to (1), in this logic tagFilter should be `Aprime` but does that match users mental model?

It's a good question -- one that I thought about quite a bit. The consequence of this behavior is that back and forward navigation within a namespace basically only adds and removes experiments. Back and forward navigation within a namespace does not change things like tagFilter and pinned cards. (On the other hand, back and forward navigation to a different namespace does change tagFilter and pinned cards.) 

Does this match the user's mental model? I'm not certain but I think it is similar to behaviors which user already experiences in tb. Consider following scenario:

(0) User loads experiment list.
(1) User chooses to view an experiment A.
(2) User sets tag filter to "filterA".
(3) User sets tag filter to "filterAprime".
(4) User sets tag filter to "filterAprimeprime".

What happens when user presses the back button? They go to the experiment list. What happens when they then press the forward button? They go to experiment A with tag filter set to "filterAprimeprime". Changes to the filter are not honored in browser history. I try to rationalize that the behavior we're building here is very similar.

I think this is a question worth asking the team when we do the bug hunt / teamfood and see if we should adjust. Would that work for you?

File: tensorboard/webapp/app_routing/effects/app_routing_effects_test.ts [hunk 1]
```diff
@@ -773,6 +774,69 @@ describe('app_routing_effects', () => {
...
+              params: {experimentIds: 'a:b'},
+              pathname: '/compare/a:b',
+              queryParams: [],
+            } as unknown as Route),
+          }),
+          jasmine.any(Object), // actions.navigated
+        ]);
+      }));
+
+      it('does not dispatch stateRehydratedFromUrl when unknown namespace id', fakeAsync(() => {
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/webapp/app_routing/effects/app_routing_effects_test.ts]: Whoops, had "unknown"/"known" flipped in this test and the one before. Thanks for catching.

**Review by japie1235813 — APPROVED** [review]
The example you mentioned resolves my confusions. 

Let me know if the following understanding is correct:
(1) and (2) are in the same namespace A and in that namespace the tagFilter has been set to `Aprime`. "A" is not remembered anywhere. Yet in the browser history it would be `urlA` and `urlAprime`. Users can hit back button twice to urlA but tagFilter has been set to `Aprime`

Also I rethink my question after trying filter changing in tb.corp, every character change triggers url change, the "old" filters are no longer existed so nor would they be in the browser history. I think the current flow somewhat matches current users' mental model. Thanks for the clarification. 
We can also see what the team suggestion in bug hunt but now I think we are on the same page.

**Comment by bmd3k** [comment]
> The example you mentioned resolves my confusions.
> 
> Let me know if the following understanding is correct: (1) and (2) are in the same namespace A and in that namespace the tagFilter has been set to `Aprime`. "A" is not remembered anywhere. Yet in the browser history it would be `urlA` and `urlAprime`. Users can hit back button twice to urlA but tagFilter has been set to `Aprime`
>

Yes, correct.
 
> Also I rethink my question after trying filter changing in tb.corp, every character change triggers url change, the "old" filters are no longer existed so nor would they be in the browser history. I think the current flow somewhat matches current users' mental model. Thanks for the clarification. We can also see what the team suggestion in bug hunt but now I think we are on the same page.

I agree, I think we have the same understanding. Great!

**Pull request merged**